tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,279 @@
+
+
+class SplunkHecDelivery(object):
+    """
+    Delivery class to send c7n message from SQS to Splunk HTTP Event Collector
+    """
+
+    def __init__(self, config, session, logger):
+        """
+        Initialize SplunkHecDelivery HEC sender.
+
+        :param config: sqs_splunk_lambda configuration
+        :type config: dict
+        :param session: boto3 AWS Session
+        :param logger: Logger object to write to
+        """
+        self.config = config
+        self.logger = logger
+        self.session = session
+
+    def get_splunk_payloads(self, msg, msg_timestamp):
+        """
+        Given a raw c7n message dict, generate a list of payloads to send to the
+        Splunk HEC.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :param msg_timestamp: int timestamp (milliseconds) when message was sent
+          to SQS, for use as event timestamp in Splunk.
+        :type msg_timestamp: int
+        :return: list of Splunk payload dicts
+        :rtype: list
+        """
+        payloads = []
+        events = self.get_splunk_events(msg)
+        indices = self._splunk_indices_for_message(msg)
+        for event in events:
+            for index in indices:
+                payloads.append({
+                    'time': msg_timestamp,
+                    'host': 'cloud-custodian',
+                    'source': '%s-cloud-custodian' % event.get(
+                        'account', 'unknown'
+                    ),
+                    'sourcetype': '_json',
+                    'index': index,
+                    'event': event
+                })
+        return payloads
+
+    def get_splunk_events(self, msg):
+        """
+        Given a raw c7n message dict, generate a list of per-resource event
+        dictionaries to send to Splunk as events.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :return: list of Splunk HEC Event dicts
+        :rtype: list
+        """
+        # if an event is present, add it to the log
+        if msg.get('event', None) is not None:
+            # add user to the log message, at the top level...
+            user = get_aws_username_from_event(self.logger, msg['event'])
+            if user is not None:
+                msg['event_triggering_user'] = user
+        # get a copy of the message with no resources
+        base_log = deepcopy(msg)
+        del base_log['resources']
+        # if configured, build and add actions list
+        if self.config.get('splunk_actions_list', False):
+            base_log['actions'] = []
+            for a in msg['policy']['actions']:
+                if isinstance(a, type({})):
+                    base_log['actions'].append(a['type'])
+                else:
+                    base_log['actions'].append(a)
+        # generate a separate Splunk message for each resource
+        logs = []
+        for res in msg['resources']:
+            x = deepcopy(base_log)
```

Comment:
afaics you can get away with a shallow copy here (x = dict(base_log), since its just adding the resource key in.